# IM939 Lab 2 - Part 1

Last session we loaded data using [Pandas](https://pandas.pydata.org/). Here we explore how to use Pandas to read in, process and explore data.

As before we load pandas and use the read_csv method.

In [1]:
import pandas as pd

df = pd.read_csv('data/raw/office_ratings.csv', encoding='UTF-8')

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188 entries, 0 to 187
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   season       188 non-null    int64  
 1   episode      188 non-null    int64  
 2   title        188 non-null    object 
 3   imdb_rating  188 non-null    float64
 4   total_votes  188 non-null    int64  
 5   air_date     188 non-null    object 
dtypes: float64(1), int64(3), object(2)
memory usage: 8.9+ KB


## Help!

Python has inbuilt documentation. To access this add a ? before an object or method.

For example, our dataframe

In [3]:
?df.info

Signature:
df.info(
    verbose: 'bool | None' = None,
    buf: 'WriteBuffer[str] | None' = None,
    max_cols: 'int | None' = None,
    memory_usage: 'bool | str | None' = None,
    show_counts: 'bool | None' = None,
) -> 'None'
Docstring:
Print a concise summary of a DataFrame.

This method prints information about a DataFrame including
the index dtype and columns, non-null values and memory usage.

Parameters
----------
verbose : bool, optional
    Whether to print the full summary. By default, the setting in
    ``pandas.options.display.max_info_columns`` is followed.
buf : writable buffer, defaults to sys.stdout
    Where to send the output. By default, the output is printed to
    sys.stdout. Pass a writable buffer if you need to further process
    the output.
max_cols : int, optional
    When to switch from the verbose to the truncated output. If the
    DataFrame has more than `max_cols` columns, the truncated output
    is used. By default, the setting in
    ``pandas.options

The dtypes property (properties of obejct are values associated with the object and are not called with a () at the end).

In [4]:
?df.dtypes

Type:        property
String form: <property object at 0x110093ce0>
Docstring:  
Return the dtypes in the DataFrame.

This returns a Series with the data type of each column.
The result's index is the original DataFrame's columns. Columns
with mixed types are stored with the ``object`` dtype. See
:ref:`the User Guide <basics.dtypes>` for more.

Returns
-------
pandas.Series
    The data type of each column.

Examples
--------
>>> df = pd.DataFrame({'float': [1.0],
...                    'int': [1],
...                    'datetime': [pd.Timestamp('20180310')],
...                    'string': ['foo']})
>>> df.dtypes
float              float64
int                  int64
datetime    datetime64[ns]
string              object
dtype: object

The info method for dataframes.

In [5]:
?df.info

Signature:
df.info(
    verbose: 'bool | None' = None,
    buf: 'WriteBuffer[str] | None' = None,
    max_cols: 'int | None' = None,
    memory_usage: 'bool | str | None' = None,
    show_counts: 'bool | None' = None,
) -> 'None'
Docstring:
Print a concise summary of a DataFrame.

This method prints information about a DataFrame including
the index dtype and columns, non-null values and memory usage.

Parameters
----------
verbose : bool, optional
    Whether to print the full summary. By default, the setting in
    ``pandas.options.display.max_info_columns`` is followed.
buf : writable buffer, defaults to sys.stdout
    Where to send the output. By default, the output is printed to
    sys.stdout. Pass a writable buffer if you need to further process
    the output.
max_cols : int, optional
    When to switch from the verbose to the truncated output. If the
    DataFrame has more than `max_cols` columns, the truncated output
    is used. By default, the setting in
    ``pandas.options

The below is quite long. But goes give you the various arguments (options) you can use with the method.

In [6]:
?pd.read_csv

Signature:
pd.read_csv(
    filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]',
    *,
    sep: 'str | None | lib.NoDefault' = <no_default>,
    delimiter: 'str | None | lib.NoDefault' = None,
    header: "int | Sequence[int] | None | Literal['infer']" = 'infer',
    names: 'Sequence[Hashable] | None | lib.NoDefault' = <no_default>,
    index_col: 'IndexLabel | Literal[False] | None' = None,
    usecols=None,
    dtype: 'DtypeArg | None' = None,
    engine: 'CSVEngine | None' = None,
    converters=None,
    true_values=None,
    false_values=None,
    skipinitialspace: 'bool' = False,
    skiprows=None,
    skipfooter: 'int' = 0,
    nrows: 'int | None' = None,
    na_values=None,
    keep_default_na: 'bool' = True,
    na_filter: 'bool' = True,
    verbose: 'bool' = False,
    skip_blank_lines: 'bool' = True,
    parse_dates: 'bool | Sequence[Hashable] | None' = None,
    infer_datetime_format: 'bool | lib.NoDefault' = <no_default>,
    keep_date_col: 'bool' =

The Pandas documentation is rather good. Relevent to our below work is:

* [What kind of data does pandas handle?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html#min-tut-01-tableoriented)
* [How to calculate summary statistics?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/01_table_oriented.html#min-tut-01-tableoriented)
* [How to create plots in pandas?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/04_plotting.html#min-tut-04-plotting)
* [How to handle time series data with ease?](https://pandas.pydata.org/pandas-docs/stable/getting_started/intro_tutorials/09_timeseries.html#min-tut-09-timeseries)

I also found [a rather nice series of lessons a kind person put together](https://bitbucket.org/hrojas/learn-pandas/src/master/). There are lots of online tutorials which will help you.

## Structure

How do we find out the structure of our data?

Well, the variable df is now a pandas DataFrame object.

In [7]:
type(df)

pandas.core.frame.DataFrame

The DataFrame object has lots of built in methods and attributes.

The info method gives us information about datatypes, dimensions and the presence of null values in our dataframe.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188 entries, 0 to 187
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   season       188 non-null    int64  
 1   episode      188 non-null    int64  
 2   title        188 non-null    object 
 3   imdb_rating  188 non-null    float64
 4   total_votes  188 non-null    int64  
 5   air_date     188 non-null    object 
dtypes: float64(1), int64(3), object(2)
memory usage: 8.9+ KB


We can just look at the datatypes if we want.

In [9]:
df.dtypes

season           int64
episode          int64
title           object
imdb_rating    float64
total_votes      int64
air_date        object
dtype: object

Or just the dimensions.

In [10]:
df.shape

(188, 6)

In this case, there are only 188 rows. But for larger datasets we might want to look at the head (top 5) and tail (bottom 5) rows.

In [11]:
df.head()

season  episode          title  imdb_rating  total_votes    air_date
0       1        1          Pilot          7.6         3706  2005-03-24
1       1        2  Diversity Day          8.3         3566  2005-03-29
2       1        3    Health Care          7.9         2983  2005-04-05
3       1        4   The Alliance          8.1         2886  2005-04-12
4       1        5     Basketball          8.4         3179  2005-04-19

In [12]:
df.tail()

season  episode             title  imdb_rating  total_votes    air_date
183       9       19     Stairmageddon          8.0         1484  2013-04-11
184       9       20    Paper Airplane          8.0         1482  2013-04-25
185       9       21  Livin' the Dream          8.9         2041  2013-05-02
186       9       22          A.A.R.M.          9.3         2860  2013-05-09
187       9       23            Finale          9.7         7934  2013-05-16

We may need to put together a dataset.

Consider the two dataframe below

In [13]:
df_1 = pd.read_csv('data/raw/office1.csv', encoding='UTF-8')
df_2 = pd.read_csv('data/raw/office2.csv', encoding='UTF-8')

In [14]:
df_1.head()

id  season  episode  imdb_rating
0   5-1       5        1          8.8
1  9-13       9       13          7.7
2   5-6       5        6          8.5
3  3-23       3       23          9.3
4  9-16       9       16          8.2

In [15]:
df_2.head()

id  total_votes
0  4-10         2095
1  3-21         2403
2  7-24         2040
3  6-18         1769
4   8-8         1584

The total votes and imdb ratings data are split between files. There is a common column called id. We can join the two dataframes together using the common column.

In [16]:
inner_join_office_df = pd.merge(df_1, df_2, on='id', how='inner')
inner_join_office_df

id  season  episode  imdb_rating  total_votes
0     5-1       5        1          8.8         2501
1    9-13       9       13          7.7         1394
2     5-6       5        6          8.5         2018
3    3-23       3       23          9.3         3010
4    9-16       9       16          8.2         1572
..    ...     ...      ...          ...          ...
183  5-21       5       21          8.7         2032
184  2-13       2       13          8.3         2363
185   9-6       9        6          7.8         1455
186   2-2       2        2          8.2         2736
187   3-4       3        4          8.0         2311

[188 rows x 5 columns]

In this way you can combine datasets using common columns. We will leave that for the moment. If you want more information about merging data then see [this page](https://www.datasciencemadesimple.com/join-merge-data-frames-pandas-python/#:~:text=Merge%20%28%29%20Function%20in%20pandas%20is%20similar%20to,rows%20from%20both%20data%20frames%2C%20specify%20how%3D%20%E2%80%98outer%E2%80%99.) and the [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html).

## Summary

To get an overview of our data we can ask Python to 'describe our data'

In [17]:
df.describe()

season     episode  imdb_rating  total_votes
count  188.000000  188.000000   188.000000   188.000000
mean     5.468085   11.877660     8.257447  2126.648936
std      2.386245    7.024855     0.538067   787.098275
min      1.000000    1.000000     6.700000  1393.000000
25%      3.000000    6.000000     7.900000  1631.500000
50%      6.000000   11.500000     8.200000  1952.500000
75%      7.250000   18.000000     8.600000  2379.000000
max      9.000000   26.000000     9.700000  7934.000000

or we can pull out specific statistics.

In [18]:
df.mean(numeric_only=True)

season            5.468085
episode          11.877660
imdb_rating       8.257447
total_votes    2126.648936
dtype: float64

Note the error triggered above due to pandas attempting to calculate the mean of the wrong type (i.e. non-numeric values). We can address that by only computing the mean of numeric values (see below):

In [19]:
df.mean(numeric_only=True)

season            5.468085
episode          11.877660
imdb_rating       8.257447
total_votes    2126.648936
dtype: float64

or the sum.

In [20]:
df.sum()

season                                                      1028
episode                                                     2233
title          PilotDiversity DayHealth CareThe AllianceBaske...
imdb_rating                                               1552.4
total_votes                                               399810
air_date       2005-03-242005-03-292005-04-052005-04-122005-0...
dtype: object

Similarly to what happened with `mean()`, `sum()` is adding all values in every observation of every attribute, regardless of their type. Can you see what happens with strings? And with dates?

Again, we can only force to use numeric values only:

In [21]:
df.sum(numeric_only=True)

season           1028.0
episode          2233.0
imdb_rating      1552.4
total_votes    399810.0
dtype: float64

Calculating these statistics for specific columns is straight forward.

In [22]:
df['imdb_rating'].mean()

8.25744680851064

In [23]:
df['total_votes'].sum()

399810

## Selecting subsets

We can even select more than one column!

In [24]:
df[['imdb_rating', 'total_votes']].mean()

imdb_rating       8.257447
total_votes    2126.648936
dtype: float64

Two sets of squared brackets is needed because you are passing a list of the column names to the getitem dunder method of the pandas dataframe object (thank [this stackoverflow question](https://stackoverflow.com/questions/11285613/selecting-multiple-columns-in-a-pandas-dataframe)). You can also check out the pandas documentation on [indexing and selecting data](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#basics).

You can also select by row and column name using the iloc method. You can specify the [row, column]. So to choose the value in the 2nd row and 4th column.

In [25]:
df.iloc[2,4]

2983

All the rows or all the columns are indicated by :. Such as,

In [26]:
df.iloc[:,2]

0                 Pilot
1         Diversity Day
2           Health Care
3          The Alliance
4            Basketball
             ...       
183       Stairmageddon
184      Paper Airplane
185    Livin' the Dream
186            A.A.R.M.
187              Finale
Name: title, Length: 188, dtype: object

In [27]:
df.iloc[2,:]

season                   1
episode                  3
title          Health Care
imdb_rating            7.9
total_votes           2983
air_date        2005-04-05
Name: 2, dtype: object

We can use negative values in indexes to indicate 'from the end'. So, an index of [-10, :] returns the 10th from last row.

In [28]:
df.iloc[-10,:]

season                  9
episode                14
title           Vandalism
imdb_rating           7.6
total_votes          1402
air_date       2013-01-31
Name: 178, dtype: object

Instead of using tail, we could ask for the last 10 rows with an index of [-10:, :]. I read : as 'and everything else' in these cases.

In [29]:
df.iloc[-5:,:]

season  episode             title  imdb_rating  total_votes    air_date
183       9       19     Stairmageddon          8.0         1484  2013-04-11
184       9       20    Paper Airplane          8.0         1482  2013-04-25
185       9       21  Livin' the Dream          8.9         2041  2013-05-02
186       9       22          A.A.R.M.          9.3         2860  2013-05-09
187       9       23            Finale          9.7         7934  2013-05-16

In [30]:
df.tail()

season  episode             title  imdb_rating  total_votes    air_date
183       9       19     Stairmageddon          8.0         1484  2013-04-11
184       9       20    Paper Airplane          8.0         1482  2013-04-25
185       9       21  Livin' the Dream          8.9         2041  2013-05-02
186       9       22          A.A.R.M.          9.3         2860  2013-05-09
187       9       23            Finale          9.7         7934  2013-05-16

Note that the row is shown on the left. That will stop you getting lost in slices of the data. 

For the top ten rows

In [31]:
df.iloc[:10,:]

season  episode              title  imdb_rating  total_votes    air_date
0       1        1              Pilot          7.6         3706  2005-03-24
1       1        2      Diversity Day          8.3         3566  2005-03-29
2       1        3        Health Care          7.9         2983  2005-04-05
3       1        4       The Alliance          8.1         2886  2005-04-12
4       1        5         Basketball          8.4         3179  2005-04-19
5       1        6           Hot Girl          7.8         2852  2005-04-26
6       2        1        The Dundies          8.7         3213  2005-09-20
7       2        2  Sexual Harassment          8.2         2736  2005-09-27
8       2        3    Office Olympics          8.4         2742  2005-10-04
9       2        4           The Fire          8.4         2713  2005-10-11

Of course, we can run methods on these slices. We could, if we wanted to, calculate the mean imdb rating of only the first and last 100 episodes. _Note_ the indexing starts at 0 so we want the column index of 3 (0:season, 1:episode, 2:title, 3:imdb_rating).

In [32]:
df.iloc[:100,3].mean()

8.483

In [33]:
df.iloc[-100:,3].mean()

8.062

If you are unsure how many rows you have then the count method comes to the rescue.

In [34]:
df.iloc[-100:,3].count()

100

In [35]:
df.describe()

season     episode  imdb_rating  total_votes
count  188.000000  188.000000   188.000000   188.000000
mean     5.468085   11.877660     8.257447  2126.648936
std      2.386245    7.024855     0.538067   787.098275
min      1.000000    1.000000     6.700000  1393.000000
25%      3.000000    6.000000     7.900000  1631.500000
50%      6.000000   11.500000     8.200000  1952.500000
75%      7.250000   18.000000     8.600000  2379.000000
max      9.000000   26.000000     9.700000  7934.000000

So it looks like the last 100 episodes were less good than the first 100. I guess that is why it was cancelled.

Our data is organised by season. Looking at the average by season might help.

In [36]:
df[['season', 'imdb_rating']].groupby('season').mean()

imdb_rating
season             
1          8.016667
2          8.436364
3          8.573913
4          8.600000
5          8.492308
6          8.219231
7          8.316667
8          7.666667
9          7.956522

The above line groups our dataframe by values in the season column and then displays the mean for each group. Pretty nifty.

Season 8 looks pretty bad. We can look at just the rows for season 8.

In [37]:
df[df['season'] == 8]

season  episode                        title  imdb_rating  total_votes  \
141       8        1                     The List          8.2         1829   
142       8        2                The Incentive          8.2         1668   
143       8        3                        Lotto          7.3         1601   
144       8        4                 Garden Party          8.1         1717   
145       8        5                      Spooked          7.6         1543   
146       8        6                     Doomsday          7.8         1476   
147       8        7            Pam's Replacement          7.7         1563   
148       8        8                   Gettysburg          7.0         1584   
149       8        9              Mrs. California          7.7         1553   
150       8       10             Christmas Wishes          8.0         1547   
151       8       11                       Trivia          7.9         1488   
152       8       12                   Pool Party          8.0         1612   
153       8       13                    Jury Duty          7.5         1478   
154       8       14              Special Project          7.8         1432   
155       8       15                  Tallahassee          7.9         1522   
156       8       16                  After Hours          8.1         1567   
157       8       17               Test the Store          7.8         1478   
158       8       18          Last Day in Florida          7.8         1429   
159       8       19                 Get the Girl          6.7         1642   
160       8       20                Welcome Party          7.2         1489   
161       8       21                   Angry Andy          7.1         1585   
162       8       22                   Fundraiser          7.1         1453   
163       8       23                     Turf War          7.7         1393   
164       8       24  Free Family Portrait Studio          7.8         1464   

       air_date  
141  2011-09-22  
142  2011-09-29  
143  2011-10-06  
144  2011-10-13  
145  2011-10-27  
146  2011-11-03  
147  2011-11-10  
148  2011-11-17  
149  2011-12-01  
150  2011-12-08  
151  2012-01-12  
152  2012-01-19  
153  2012-02-02  
154  2012-02-09  
155  2012-02-16  
156  2012-02-23  
157  2012-03-01  
158  2012-03-08  
159  2012-03-15  
160  2012-04-12  
161  2012-04-19  
162  2012-04-26  
163  2012-05-03  
164  2012-05-10

We can get an overview of the rating of all chapters within season 8 by:

In [38]:
df.loc[df['season'] == 8, 'imdb_rating'].describe()

count    24.000000
mean      7.666667
std       0.405041
min       6.700000
25%       7.450000
50%       7.800000
75%       7.925000
max       8.200000
Name: imdb_rating, dtype: float64

In [39]:
df['season'] == 8


0      False
1      False
2      False
3      False
4      False
       ...  
183    False
184    False
185    False
186    False
187    False
Name: season, Length: 188, dtype: bool

Generally pretty bad, but there is clearly one very disliked episode.

## Adding columns

We can add new columns pretty simply.

In [40]:
df['x'] = 44
df.head()

season  episode          title  imdb_rating  total_votes    air_date   x
0       1        1          Pilot          7.6         3706  2005-03-24  44
1       1        2  Diversity Day          8.3         3566  2005-03-29  44
2       1        3    Health Care          7.9         2983  2005-04-05  44
3       1        4   The Alliance          8.1         2886  2005-04-12  44
4       1        5     Basketball          8.4         3179  2005-04-19  44

Our new column can be an operation on other columns

In [41]:
df['rating_div_total_votes'] = df['imdb_rating'] / df['total_votes']
df.head()

season  episode          title  imdb_rating  total_votes    air_date   x  \
0       1        1          Pilot          7.6         3706  2005-03-24  44   
1       1        2  Diversity Day          8.3         3566  2005-03-29  44   
2       1        3    Health Care          7.9         2983  2005-04-05  44   
3       1        4   The Alliance          8.1         2886  2005-04-12  44   
4       1        5     Basketball          8.4         3179  2005-04-19  44   

   rating_div_total_votes  
0                0.002051  
1                0.002328  
2                0.002648  
3                0.002807  
4                0.002642

or as simple as adding one to every value.

In [42]:
df['y'] = df['season'] + 1
df.iloc[0:5,:]

season  episode          title  imdb_rating  total_votes    air_date   x  \
0       1        1          Pilot          7.6         3706  2005-03-24  44   
1       1        2  Diversity Day          8.3         3566  2005-03-29  44   
2       1        3    Health Care          7.9         2983  2005-04-05  44   
3       1        4   The Alliance          8.1         2886  2005-04-12  44   
4       1        5     Basketball          8.4         3179  2005-04-19  44   

   rating_div_total_votes  y  
0                0.002051  2  
1                0.002328  2  
2                0.002648  2  
3                0.002807  2  
4                0.002642  2

In [43]:
y =  df['season'] + 1

## Writing data

Pandas supports writing out data frames to various formats.

In [44]:
?df.to_csv

Signature:
df.to_csv(
    path_or_buf: 'FilePath | WriteBuffer[bytes] | WriteBuffer[str] | None' = None,
    sep: 'str' = ',',
    na_rep: 'str' = '',
    float_format: 'str | Callable | None' = None,
    columns: 'Sequence[Hashable] | None' = None,
    header: 'bool_t | list[str]' = True,
    index: 'bool_t' = True,
    index_label: 'IndexLabel | None' = None,
    mode: 'str' = 'w',
    encoding: 'str | None' = None,
    compression: 'CompressionOptions' = 'infer',
    quoting: 'int | None' = None,
    quotechar: 'str' = '"',
    lineterminator: 'str | None' = None,
    chunksize: 'int | None' = None,
    date_format: 'str | None' = None,
    doublequote: 'bool_t' = True,
    escapechar: 'str | None' = None,
    decimal: 'str' = '.',
    errors: 'str' = 'strict',
    storage_options: 'StorageOptions' = None,
) -> 'str | None'
Docstring:
Write object to a comma-separated values (csv) file.

Parameters
----------
path_or_buf : str, path object, file-like object, or None, default None
  

Now you can uncomment the code below to save your dataframe into a csv file. But before doing so, check that your `data/output` folder is empty:

In [45]:
# df.to_csv('data/output/my_output_ratings.csv', encoding='UTF-8')

In [46]:
?df.to_excel

Signature:
df.to_excel(
    excel_writer,
    sheet_name: 'str' = 'Sheet1',
    na_rep: 'str' = '',
    float_format: 'str | None' = None,
    columns: 'Sequence[Hashable] | None' = None,
    header: 'Sequence[Hashable] | bool_t' = True,
    index: 'bool_t' = True,
    index_label: 'IndexLabel' = None,
    startrow: 'int' = 0,
    startcol: 'int' = 0,
    engine: 'str | None' = None,
    merge_cells: 'bool_t' = True,
    inf_rep: 'str' = 'inf',
    freeze_panes: 'tuple[int, int] | None' = None,
    storage_options: 'StorageOptions' = None,
) -> 'None'
Docstring:
Write object to an Excel sheet.

To write a single object to an Excel .xlsx file it is only necessary to
specify a target file name. To write to multiple sheets it is necessary to
create an `ExcelWriter` object with a target file name, and specify a sheet
in the file to write to.

Multiple sheets may be written to by specifying unique `sheet_name`.
With all data written to the file it is necessary to save the changes.
Note that

Now you can uncomment the code below to save your dataframe into an excel file. But before doing so, check that your `data/output` folder is empty:

In [47]:
# df.to_excel('data/output/my_output_ratings.xlsx')